pom.xml

```diff
@@ -845,5 +841,23 @@ Import-Package: \\
         </pluginManagement>
       </build>
     </profile>
+    <profile>
+      <id>j17</id>
+      <activation>
+        <jdk>[17.0,18.0)</jdk>
+      </activation>
+      <properties>
+        <maven.compiler.release>17</maven.compiler.release>
+      </properties>
+    </profile>
+    <profile>
+      <id>j21</id>
+      <activation>
+        <jdk>[21.0,22.0)</jdk>
+      </activation>
+      <properties>
+        <maven.compiler.release>21</maven.compiler.release>
```

Comment:
It's probably better to keep the artifacts compatible with Java 17 regardless of the installed JDK so the artifacts remain compatible with the official release. That way you can use Java 21 on other projects without having to switch JDKs when working on openHAB. It will also help with preventing contributors using newer Java constructs that aren't available on Java 17. They would then only get this feedback when CI fails.

Reply:
@wborn Thanks for looking into this. Keeping release to 17 basically means that the JDK could be 21, but still behaving like J17. It addresses the point that users have a JDK21 installed. OH modules would then still be J17.

I personally had a different use case in mind, "compiling and running" on J21.
The build anyway results in a mixture of different class file versions. I would try to upgrade my test server to JDK21, thus J21 class files should be fine. I did not have the use case above in mind, this is why I proposed the auto-detection:
https://github.com/openhab/openhab-distro/issues/1590#issuecomment-2019099635
You are right, only CI will fail when users start to use new J21 features in this case. I do not have a solution for this.

Do you think it is a better approach to use release set to 17 unless a profile for J21 is explicitly activated?

The question is more or less about the approach we are taking towards an upcoming OH5.x which might require J21. How do we gain knowledge about potential problems when compiling and running in J21.....

Reply:
We can add a GHA PR build for J21 like we did in the past. In such a build it can activate the profile to use J21 release.

OH5 will probably still take some. We just started using J17 with OH4. Maybe an OH5 will happen after J25 is released in 2025. 🔮 😉

The issue with requiring newer Java versions is that users may need to manually upgrade their JDK, OS or even hardware like we saw after requiring J17.

Running OH4 using J21 probably already works very well with all the recent updates.

Reply:
I changed the POM file, now default it profile ```j17``` for both JDK 17 and JDK 21.
Profile ```j21``` can be activated explicitly to build native Java 21 file.